### PR TITLE
feat: Introduce hamming code generator

### PR DESCRIPTION
**Pull request by AyushChakraborty**
Introduced a new .py file which generates 16 bit
hamming codes from the given 11 bit input.
Additional 5 bits are redundant, used to correct
single bit errors within the data bits

### Describe your change:
Introduced a new .py file called hamming_code_generator.py which generates 16 bit hamming codes. 5 bits introduced are for redundancy and used to check and correct for a single bit change anywhere in the final 16 bit binary number. Doctests are written and all other rules mentioned in the checklist and README.md file are followed by me.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Add or change doctests? -- Note: Please avoid changing both code and tests in a single pull request.
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by AyushChakraborty** [comment]
made the lines in the docstring shorter, imported modules according to the specified norm and added k values to a list and used membership operation instead of checking the values of k individually in one of the if statements

File: bit_manipulation/hamming_code_generator.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    >>> hamming_15_11("10111100110")
+    '0011001101100110'
+    >>> hamming_15_11("10001000010")
+    '0001100001000010'
+
+    """
+    if len(number) == 11:
+        digits = [0 if number[i] == "0" else 1 for i in range(len(number))]
+        total_num_1 = sum(digits)
+        hamming_digits = [0 for i in range(16)]
```
Comment (jeffreyyancey) [inline bit_manipulation/hamming_code_generator.py:41]: Cool approach. I feel like the reduce + lambda can be hard to follow along at times. I believe the following has the same result and is more intuitive to follow:

```
parity_positions = {
    1: [0, 1, 3, 4, 6, 8, 10],
    2: [0, 2, 3, 5, 6, 9, 10],
    4: [1, 2, 3, 7, 8, 9, 10],
    8: [4, 5, 6, 7, 8, 9, 10],
}

parity_bits = {}
for parity_pos, data_indices in parity_positions.items():
    parity = 0
    for idx in data_indices:
        parity ^= digits[idx]
    parity_bits[parity_pos] = parity
```

File: bit_manipulation/hamming_code_generator.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
+""" "
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:1]: Oh yes, will remove it

File: bit_manipulation/hamming_code_generator.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+Author: Ayush Chakraborty
+Date: 6th October 2024
+
+generate (15,11) hamming encoded bits gives 11 bit data
+
+input: 11 bit binary number with type string
+return: 16 bit binary hamming encoded number returned in string form
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:8]: Actually no as the final number returned will be 16 bits, the 1 extra bit is for the final check of all the 15 blocks, where its just a XOR of all of them

File: bit_manipulation/hamming_code_generator.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    Performs parity checks to assign values to the redundant bits,
+    in the 16 bit number
+    returned, bits at index 0, 1, 2, 4, 8 are redundant bits used for checking
+
+    Hamming generated 16 bits generated from the 11 bit binary number can only detect
+    and change a single bit change, but can only detect more than a
+    single bit change
+
+    for more theoretical knowledege about Hamming codes, refer:
+    https://www.youtube.com/watch?v=X8jsijhllIA&t=143s
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:25]: Oh yes, its not really relevant, will correct it to the start, and also will add the wiki

File: bit_manipulation/hamming_code_generator.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    '0010101110001110'
+    >>> hamming_15_11("10110000110")
+    '0101001100000110'
+    >>> hamming_15_11("10111100110")
+    '0011001101100110'
+    >>> hamming_15_11("10001000010")
+    '0001100001000010'
+
+    """
+    if len(number) == 11:
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:38]: Ah yes, this check. makes sense too, will add

File: bit_manipulation/hamming_code_generator.py (hunk 1 above)
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:41]: Thanks! Introducing similar changes

File: bit_manipulation/hamming_code_generator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:39]: I added a else statement instead so that it can take care of all other cases where input is not a 11 bit number (int str format)

**Comment by AyushChakraborty** [comment]
@jeffreyyancey added the appropriate changes

**Comment by jeffreyyancey** [comment]
The only concern I have left is that it doesn't appear your are enforcing that the `number` that comes in contains only "0" or "1" characters. It appears to just be converting any character that isn't a "0" to 1. So a string like "#&! & ((30qhjj1+" would be accepted and become 1111111110111111.

**Comment by AyushChakraborty** [comment]
> The only concern I have left is that it doesn't appear your are enforcing that the `number` that comes in contains only "0" or "1" characters. It appears to just be converting any character that isn't a "0" to 1. So a string like "#&! & ((30qhjj1+" would be accepted and become 1111111110111111.

Oh yes, I completely overlooked this, thanks I'll make the changes.

**Review by jeffreyyancey — APPROVED** [review]

**Comment by AyushChakraborty** [comment]
Updated the code, can you, if possible merge the PR now?

File: bit_manipulation/hamming_code_generator.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+        redundant_bits[0] = (
+            total_num_1 % 2
+            ^ redundant_bits[1]
+            ^ redundant_bits[2]
+            ^ redundant_bits[3]
+            ^ redundant_bits[4]
+        )
+
+        j = -1
```
Comment (jeffreyyancey) [inline bit_manipulation/hamming_code_generator.py:78]: I'd give this a name like `data_index` and set it to `0`, then update via `data_index += 1` after filling the hamming_digits instead of before.

**Review by jeffreyyancey — COMMENTED** [review]
Everything looks like it is working well, just a couple minor things and I'd be happy to approve.

Primarily, can you remove `r` since it is no longer used in this implementation, and can you update the couple variable that need more descriptive names.

File: bit_manipulation/hamming_code_generator.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    is_bin = True  # assuming it's binary initially
+    for i in number:
+        if i not in ("0", "1"):
+            is_bin = False
+            break
+
+    if len(number) == 11 and is_bin:
+        digits = [int(number[i]) for i in range(len(number))]
+
+        total_num_1 = sum(digits)
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:50]: Will change this

File: bit_manipulation/hamming_code_generator.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+        redundant_bits[0] = (
+            total_num_1 % 2
+            ^ redundant_bits[1]
+            ^ redundant_bits[2]
+            ^ redundant_bits[3]
+            ^ redundant_bits[4]
+        )
+
+        j = -1
+        r = 0
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:79]: Oh, but if r is removed then line 84 and 85 where 
                hamming_digits[k] = redundant_bits[r]
                r += 1
is written wont work right?

File: bit_manipulation/hamming_code_generator.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    '0101001100000110'
+    >>> hamming_15_11("10111100110")
+    '0011001101100110'
+    >>> hamming_15_11("10001000010")
+    '0001100001000010'
+    >>> hamming_15_11("00010abcdef")
+    "Input must be an 11-bit binary string containing only '0's and '1's."
+
+    """
+    is_bin = True  # assuming it's binary initially
```
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:41]: This makes sense but the only issue with this is that I fail the pre-commit.ci - pr checks when I do this, as I dont have an else statement and that too would need a return statement. So could i also mention
           else:
                  return None
?

[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:41]: Adding an additional else worked overall and for the checks

File: bit_manipulation/hamming_code_generator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:79]: Here I could introduce another dictionary {0:0, 1:1, 2:2. 3:4, 4: 8} and that way we wont need r but is it any better to do so?

[the remark this answers is not in the document]
Comment (jeffreyyancey) [reply bit_manipulation/hamming_code_generator.py:79]: I was thinking of using `hamming_digits[k] = redundant_bits.pop(0)`, though `r` to do it yourself is fine, just git it a name and either approach works.

[the remark this answers is not in the document]
Comment (AyushChakraborty) [reply bit_manipulation/hamming_code_generator.py:79]: oh yes, its a clever method, will implement it

**Comment by AyushChakraborty** [comment]
@jeffreyyancey implemented the pop method

**Review by jeffreyyancey — APPROVED** [review]
Looks good to me.

**Comment by AyushChakraborty** [comment]
@jeffreyyancey this is actually my first open source contribution, so do you have any idea how the PR can be accepeted into the main repo?

**Comment by jeffreyyancey** [comment]
I believe that one of the admins needs to give their approval. Approvals from other users will not allow merging. Instead, it helps find any issues prior to one of the individuals with the rights to approve for merging to view.

**Comment by AyushChakraborty** [comment]
Ok got it, thanks for the help

File: bit_manipulation/hamming_code_generator.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    for i in number:
+        if i not in ("0", "1"):
+            is_bin = False
+            break
+
+    if len(number) == 11 and is_bin:
+        digits = [int(number[i]) for i in range(len(number))]
```
Comment (cclauss) [inline bit_manipulation/hamming_code_generator.py:48]:
```suggestion
        digits = [int(digit) for digit in number]
```

File: bit_manipulation/hamming_code_generator.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+                hamming_digits[k] = digits[data_index]
+                data_index += 1
+
+        return "".join([str(i) for i in hamming_digits])
+
+    elif len(number) != 11 or not is_bin:
+        return "Input must be an 11-bit binary string containing only '0's and '1's."
+
+    else:
+        return "Invalid input"
```
Comment (cclauss) [inline bit_manipulation/hamming_code_generator.py:93]: These must be raised exceptions.

**Comment by cclauss** [comment]
Do we need more hamming?
https://github.com/search?q=repo%3ATheAlgorithms%2FPython+hamming
